dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -0,0 +1,212 @@
+
+        private StickySessionSupplier(EndpointChannelFactory channelFactory) {
+            this.channelFactory = channelFactory;
+        }
+
+        @Override
+        public Channel get() {
+            return new Sticky(channelFactory, router);
+        }
+    }
+
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router;
+        private final RoutingKey routingKey;
+
+        private Sticky(EndpointChannelFactory channelFactory, StickyRouter router) {
+            this.router = router;
+            this.channelFactory = channelFactory;
+            this.routingKey = RoutingKey.create();
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            return new StickyEndpointChannel(router, channelFactory.endpoint(endpoint), routingKey);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+            return endpoint(endpoint).execute(request);
+        }
+
+        @Override
+        public String toString() {
+            return "Sticky{" + channelFactory + '}';
+        }
+    }
+
+    public static StickyEndpointChannels2 create(EndpointChannelFactory endpointChannelFactory) {
+        return new StickyEndpointChannels2(endpointChannelFactory);
+    }
+
+    interface StickyRouter {
+        ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel);
+    }
+
+    private static final class DefaultStickyRouter implements StickyRouter {
+
+        private final InFlightCallCallback callback = new InFlightCallCallback();
+
+        @Nullable
+        private volatile HostId hostId;
+
+        @Nullable
+        private volatile ListenableFuture<Response> callInFlight;
+
+        @Override
+        public ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel) {
+            if (hostId != null) {
+                return executeWithHostId(hostId, request, endpointChannel);
+            }
+
+            synchronized (this) {
+                if (hostId != null) {
+                    request.attachments().put(RoutingAttachments.EXECUTE_ON_HOST_ID_KEY, hostId);
+                    return endpointChannel.execute(request);
+                }
+
+                if (callInFlight == null) {
+                    callInFlight = DialogueFutures.addDirectCallback(
+                            executeWithAttachHostId(request, endpointChannel), callback);
+                    return callInFlight;
+                } else {
+                    ListenableFuture<Response> result = callInFlight;
+                    result = DialogueFutures.transformAsync(result, _input -> execute(request, endpointChannel));
```

Comment:
we need to wrap callInFlight initially to prevent `cancel` on queued calls from causing collateral damage. That's probably best to do around setting `callInFlight` initially rather than on each subsequent request.

Reply:
Grr, I swear I had this in, but must've rushed it.

Reply:
I don't think callInFlight works, because then cancel on the returned future, would not cancel the actual work future?